roofit/roofit/src/RooLagrangianMorphFunc.cxx

```diff
@@ -0,0 +1,3558 @@
+  std::ifstream in(fname);
+  if (!in.good()) {
+    std::cerr << "unable to read file '" << fname << "'!" << std::endl;
+  }
+  MatrixT mat = readMatrixFromStreamT<MatrixT>(in);
+  in.close();
+  return mat;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// convert a TH1* param hist into the corresponding ParamSet object
+
+template <class T> inline std::map<const std::string, T> readValues(TH1 *h_pc) {
+  std::map<const std::string, T> point;
+  if (h_pc) {
+    // loop over all bins of the param_card histogram
+    for (int ibx = 1; ibx <= h_pc->GetNbinsX(); ++ibx) {
+      // read the value of one parameter
+      const std::string s_coup(h_pc->GetXaxis()->GetBinLabel(ibx));
+      double coup_val = h_pc->GetBinContent(ibx);
+      // add it to the map
+      if (!s_coup.empty()) {
+        point[s_coup] = T(coup_val);
+      }
+    }
+  }
+  return point;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// retrieve a param_hist from a certain subfolder 'name' of the file
+
+inline TH1F *getParamHist(TDirectory *file, const std::string &name,
+                          const std::string &objkey = "param_card",
+                          bool notFoundError = true) {
+  TFolder *f_tmp = dynamic_cast<TFolder *>(file->Get(name.c_str()));
+  if (!f_tmp) {
+    std::cerr << "unable to retrieve folder '" << name << "' from file '"
+              << file->GetName() << "'!" << std::endl;
+    return nullptr;
+}
+  // retrieve the histogram param_card which should live directly in the folder
+  TH1F *h_pc = dynamic_cast<TH1F *>(f_tmp->FindObject(objkey.c_str()));
+  if (h_pc) {
+    return h_pc;
+  }
+  if (notFoundError) {
+    std::cerr << "unable to retrieve " << objkey << " histogram from folder '"
+              << name << "'" << std::endl;
+  }
+  return NULL;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// retrieve a ParamSet from a certain subfolder 'name' of the file
+
+template <class T>
+inline std::map<const std::string, T>
+readValues(TDirectory *file, const std::string &name,
+           const std::string &key = "param_card", bool notFoundError = true) {
+  TH1F *h_pc = getParamHist(file, name, key, notFoundError);
+  return readValues<T>(h_pc);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// retrieve the param_hists file and return a map of the parameter values
+/// by providing a list of names, only the param_hists of those subfolders are
+/// read leaving the list empty is interpreted as meaning 'read everyting'
+
+template <class T>
+inline std::map<const std::string, std::map<const std::string, T>>
+readValues(TDirectory *f, const std::vector<std::string> &names,
+           const std::string &key = "param_card", bool notFoundError = true) {
+  std::map<const std::string, std::map<const std::string, T>> inputParameters;
```

Comment:
```suggestion
void readValues(std::map<const std::string, std::map<const std::string, T>> & inputParameters,
           TDirectory *f, const std::vector<std::string> &names,
           const std::string &key = "param_card", bool notFoundError = true) {
  inputParameters.clear();
```
We better fill the maps in-place so we don't use the copy-constructor which caused trouble with the tests on macOS.

Also, I would not make these non-performance critical function `inline`.